core/async/src/test_loop/multi_instance.rs

```diff
@@ -0,0 +1,57 @@
+use std::sync::Arc;
+
+use super::{
+    delay_sender::{CanSendWithDelay, DelaySender},
+    event_handler::LoopEventHandler,
+};
+
+/// DelaySender that attaches an index to every event and forwards it on to
+/// another DelaySender that takes a tuple of the index and the event.
+pub(crate) struct IndexedDelaySender<Event> {
+    outer: Arc<dyn CanSendWithDelay<(usize, Event)>>,
+    index: usize,
```

Comment:
At this point it seems to me that you are heading towards implementing 2^|features| structs (currently features = {Delay,Indexed}). It doesn't look sustainable, just saying.

Reply:
(1) DelaySender and LoopEventHandler are fundamental primitives of the TestLoop.

(2) "Indexed", "Widening/Narrowing" are extensions that "upcasts" DelaySenders or LoopEventHandlers.

(3) And then finally there's the conversion from DelaySender to Sender and AsyncSender (the latter is in a future PR). 

And with the way it's setup, (1) does not care about (2), and (2) does not care about (3). So there isn't really much of a exponentiation, or even multiplication, of features.